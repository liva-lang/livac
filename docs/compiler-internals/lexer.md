# Lexer (Tokenization)

The lexer converts source code into tokens using the **Logos** crate for high-performance lexical analysis.

## Location

**File**: `src/lexer.rs`

## Overview

The lexer performs:
1. **Tokenization**: Converts character stream to tokens
2. **Comment Stripping**: Removes single-line (`//`) and multi-line (`/* */`) comments
3. **Whitespace Handling**: Skips whitespace automatically
4. **Span Tracking**: Records source locations for error reporting

## Token Types

### Keywords (59 tokens)

**Control Flow**:
- `let`, `const`, `if`, `else`, `while`, `for`, `in`, `switch`, `case`, `default`
- `return`, `throw`, `try`, `catch`

**Concurrency**:
- `async`, `parallel`, `par`, `task`, `fire`, `await`
- `move`, `seq`, `vec`, `parvec`, `with`

**Data-Parallel Policies**:
- `ordered`, `chunk`, `threads`, `simdWidth`
- `prefetch`, `reduction`, `schedule`, `detect`
- `auto`, `safe`, `fast`, `static`, `dynamic`

**Error Handling**:
- `fail`

**Types**:
- `number`, `float`, `bool`, `char`, `string`, `bytes`
- `type`, `import`, `use`, `rust`, `test`

**Literals**:
- `true`, `false`

**Operators** (word-based):
- `and`, `or`, `not`, `as`

### Operators

**Arithmetic**: `+`, `-`, `*`, `/`, `%`

**Comparison**: `==`, `!=`, `<`, `<=`, `>`, `>=`

**Logical**: `&&`, `||`, `!`

**Other**: `=`, `.`, `..`, `=>`, `?`, `:`

### Delimiters

`(`, `)`, `{`, `}`, `[`, `]`, `,`, `;`

### Literals

```rust
IntLiteral(i64)         // 42, 100, 1_000_000
FloatLiteral(f64)       // 3.14, 1.5, 0.001
StringLiteral(String)   // "hello", "world"
CharLiteral(char)       // 'a', 'Z', '\n'
StringTemplate(String)  // $"Hello {name}"
```

### Identifiers

```rust
Ident(String)           // public: myVar, calculate
ProtectedIdent(String)  // protected: _helper
PrivateIdent(String)    // private: __secret
```

Visibility is determined by leading underscores:
- No prefix: Public
- `_`: Protected
- `__`: Private

## Implementation

### Using Logos

```rust
#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n\f]+")]         // Skip whitespace
#[logos(skip r"//[^\n]*")]             // Skip single-line comments
#[logos(skip r"/\*([^*]|\*[^/])*\*/")]  // Skip multi-line comments
pub enum Token {
    #[token("let")]
    Let,
    
    #[token("async")]
    Async,
    
    #[regex(r"[0-9][0-9_]*", parse_int)]
    IntLiteral(i64),
    
    // ... more tokens
}
```

### Key Features

1. **Automatic Whitespace Skipping**: Using `#[logos(skip)]`
2. **Regex Patterns**: For numbers, strings, identifiers
3. **Underscore Separators**: `1_000_000` → `1000000`
4. **String Templates**: `$"Hello {name}"` tokenized separately

### Integer Parsing

```rust
#[regex(r"[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<i64>().ok()
})]
IntLiteral(i64),
```

Supports underscores for readability: `1_000_000`

### Float Parsing

```rust
#[regex(r"[0-9][0-9_]*\.[0-9][0-9_]*", |lex| {
    lex.slice().replace('_', "").parse::<f64>().ok()
})]
FloatLiteral(f64),
```

### String Literals

```rust
#[regex(r#""([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[1..s.len()-1];  // Strip quotes
    Some(content.to_string())
})]
StringLiteral(String),
```

### String Templates

```rust
#[regex(r#"\$"([^"\\]|\\.)*""#, |lex| {
    let s = lex.slice();
    let content = &s[2..s.len()-1];  // Strip $" and "
    Some(content.to_string())
})]
StringTemplate(String),
```

## TokenWithSpan

```rust
pub struct TokenWithSpan {
    pub token: Token,
    pub span: Span,
}
```

Tracks source location for error reporting:
- `span.start`: Start byte offset
- `span.end`: End byte offset
- `span.start_position(map)`: (line, column)

## Public API

```rust
// Tokenize source code
pub fn tokenize(source: &str, file_name: &str) -> Result<Vec<TokenWithSpan>>
```

Returns `Vec<TokenWithSpan>` or `CompilerError` on invalid tokens.

## Error Handling

Lexer errors include:
- **Invalid tokens**: Unrecognized characters
- **Malformed literals**: Unclosed strings, invalid numbers
- **Location info**: Line, column, source snippet

## Summary

- **50+ Keywords**: Including all concurrency and data-parallel policy tokens
- **Logos-based**: High-performance lexing
- **Span Tracking**: For precise error messages
- **Comment Stripping**: Automatic via regex
- **Visibility-Aware**: Identifier prefixes determine visibility

**Next**: [Parser →](parser.md)
