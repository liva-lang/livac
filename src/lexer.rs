use crate::error::{CompilerError, Result};
use logos::Logos;

#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t\r\n\f]+")]
#[logos(skip r"//[^\n]*")]
#[logos(skip r"/\*([^*]|\*[^/])*\*/")]
pub enum Token {
    // Keywords
    #[token("let")]
    Let,
    #[token("const")]
    Const,
    #[token("import")]
    Import,
    #[token("use")]
    Use,
    #[token("rust")]
    Rust,
    #[token("type")]
    Type,
    #[token("test")]
    Test,
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("while")]
    While,
    #[token("for")]
    For,
    #[token("in")]
    In,
    #[token("switch")]
    Switch,
    #[token("case")]
    Case,
    #[token("default")]
    Default,
    #[token("throw")]
    Throw,
    #[token("try")]
    Try,
    #[token("catch")]
    Catch,
    #[token("return")]
    Return,
    #[token("async")]
    Async,
    #[token("parallel")]
    Parallel,
    #[token("par")]
    Par,
    #[token("task")]
    Task,
    #[token("fire")]
    Fire,
    #[token("await")]
    Await,
    #[token("move")]
    Move,
    #[token("seq")]
    Seq,
    #[token("vec")]
    Vec,
    #[token("boost")]
    Boost,
    #[token("with")]
    With,
    #[token("ordered")]
    Ordered,
    #[token("chunk")]
    Chunk,
    #[token("threads")]
    Threads,
    #[token("simdWidth")]
    SimdWidth,
    #[token("prefetch")]
    Prefetch,
    #[token("reduction")]
    Reduction,
    #[token("schedule")]
    Schedule,
    #[token("detect")]
    Detect,
    #[token("auto")]
    Auto,
    #[token("safe")]
    Safe,
    #[token("fast")]
    Fast,
    #[token("static")]
    Static,
    #[token("dynamic")]
    Dynamic,
    #[token("as")]
    As,
    #[token("true")]
    True,
    #[token("false")]
    False,

    // Logical operators (words)
    #[token("and")]
    And,
    #[token("or")]
    Or,
    #[token("not")]
    Not,

    // Type keywords
    #[token("number")]
    Number,
    #[token("float")]
    Float,
    #[token("bool")]
    Bool,
    #[token("char")]
    CharType,
    #[token("string")]
    String,
    #[token("bytes")]
    Bytes,

    // Operators
    #[token("=")]
    Assign,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Star,
    #[token("/")]
    Slash,
    #[token("%")]
    Percent,
    #[token("<")]
    Lt,
    #[token("<=")]
    Le,
    #[token(">")]
    Gt,
    #[token(">=")]
    Ge,
    #[token("==")]
    Eq,
    #[token("!=")]
    Ne,
    #[token("&&")]
    AndAnd,
    #[token("||")]
    OrOr,
    #[token("!")]
    Bang,

    // Delimiters
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,
    #[token(",")]
    Comma,
    #[token(":")]
    Colon,
    #[token(";")]
    Semicolon,
    #[token(".")]
    Dot,
    #[token("..")]
    DotDot,
    #[token("?")]
    Question,
    #[token("=>")]
    Arrow,

    // Literals
    #[regex(r"[0-9][0-9_]*", |lex| lex.slice().replace('_', "").parse::<i64>().ok())]
    IntLiteral(i64),

    #[regex(r"[0-9][0-9_]*\.[0-9][0-9_]*", |lex| lex.slice().replace('_', "").parse::<f64>().ok())]
    FloatLiteral(f64),

    #[regex(r#""([^"\\]|\\.)*""#, |lex| {
        let s = lex.slice();
        let content = &s[1..s.len()-1];
        Some(content.to_string())
    })]
    StringLiteral(String),

    #[regex(r"'([^'\\]|\\.)+'", |lex| {
        let s = lex.slice();
        let content = &s[1..s.len()-1];
        content.chars().next()
    })]
    CharLiteral(char),

    // String template
    #[regex(r#"\$"([^"\\]|\\.)*""#, |lex| {
        let s = lex.slice();
        let content = &s[2..s.len()-1];
        Some(content.to_string())
    })]
    StringTemplate(String),

    // Identifiers
    #[regex(r"__[a-zA-Z_][a-zA-Z0-9_]*", |lex| lex.slice().to_string())]
    PrivateIdent(String),

    #[regex(r"_[a-zA-Z_][a-zA-Z0-9_]*", |lex| lex.slice().to_string())]
    ProtectedIdent(String),

    #[regex(r"[a-zA-Z][a-zA-Z0-9_]*", |lex| lex.slice().to_string())]
    Ident(String),
}

#[derive(Debug, Clone)]
pub struct TokenWithSpan {
    pub token: Token,
    pub span: std::ops::Range<usize>,
}

pub fn tokenize(source: &str) -> Result<Vec<TokenWithSpan>> {
    let mut lexer = Token::lexer(source);
    let mut tokens = Vec::new();

    while let Some(result) = lexer.next() {
        match result {
            Ok(token) => {
                tokens.push(TokenWithSpan {
                    token,
                    span: lexer.span(),
                });
            }
            Err(_) => {
                let span = lexer.span();
                let line = source[..span.start].lines().count();
                let col = source[..span.start]
                    .lines()
                    .last()
                    .map(|l| l.len())
                    .unwrap_or(0);

                return Err(CompilerError::LexerError(format!(
                    "Invalid token at line {}, column {}: '{}'",
                    line, col, &source[span]
                )));
            }
        }
    }

    Ok(tokens)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_tokens() {
        let source = "let x = 10";
        let tokens = tokenize(source).unwrap();

        assert_eq!(tokens[0].token, Token::Let);
        assert_eq!(tokens[1].token, Token::Ident("x".to_string()));
        assert_eq!(tokens[2].token, Token::Assign);
        assert_eq!(tokens[3].token, Token::IntLiteral(10));
    }

    #[test]
    fn test_visibility() {
        let source = "public _protected __private";
        let tokens = tokenize(source).unwrap();

        assert_eq!(tokens[0].token, Token::Ident("public".to_string()));
        assert_eq!(
            tokens[1].token,
            Token::ProtectedIdent("_protected".to_string())
        );
        assert_eq!(
            tokens[2].token,
            Token::PrivateIdent("__private".to_string())
        );
    }

    #[test]
    fn test_logical_operators() {
        let source = "and or not && || !";
        let tokens = tokenize(source).unwrap();

        assert_eq!(tokens[0].token, Token::And);
        assert_eq!(tokens[1].token, Token::Or);
        assert_eq!(tokens[2].token, Token::Not);
        assert_eq!(tokens[3].token, Token::AndAnd);
        assert_eq!(tokens[4].token, Token::OrOr);
        assert_eq!(tokens[5].token, Token::Bang);
    }

    #[test]
    fn test_string_template() {
        let source = r#"$"Hello {name}""#;
        let tokens = tokenize(source).unwrap();

        assert!(matches!(tokens[0].token, Token::StringTemplate(_)));
    }
}
