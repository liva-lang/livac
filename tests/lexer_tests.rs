use insta::assert_snapshot;
use livac::lexer::tokenize;
use std::fs;

/// Test helper para casos correctos del lexer
fn test_lexer_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/lexer/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    
    // Convertir tokens a formato legible para snapshot
    let token_strings: Vec<String> = tokens
        .iter()
        .map(|t| format!("{:?}", t.token))
        .collect();
    
    let output = format!("Tokens: {:#?}", token_strings);
    assert_snapshot!(format!("ok_{}.tokens", test_name), output);
}

/// Test helper para casos de error del lexer
fn test_lexer_err(test_name: &str) {
    let source = fs::read_to_string(format!("tests/lexer/err_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: err_{}.liva", test_name));
    
    let result = tokenize(&source);
    assert!(result.is_err(), "Expected lexer error for test: {}", test_name);
    
    let error_msg = result.unwrap_err().to_string();
    assert_snapshot!(format!("err_{}.diag", test_name), error_msg);
}

#[test]
fn test_identifiers() {
    test_lexer_ok("identifiers");
}

#[test]
fn test_literals() {
    test_lexer_ok("literals");
}

#[test]
fn test_operators() {
    test_lexer_ok("operators");
}

#[test]
fn test_keywords() {
    test_lexer_ok("keywords");
}

#[test]
fn test_comments() {
    test_lexer_ok("comments");
}

#[test]
fn test_unknown_token() {
    test_lexer_err("unknown_token");
}

#[test]
fn test_unclosed_string() {
    test_lexer_err("unclosed_string");
}

#[test]
fn test_unclosed_char() {
    test_lexer_err("unclosed_char");
}

#[test]
fn test_unclosed_comment() {
    test_lexer_err("unclosed_comment");
}
