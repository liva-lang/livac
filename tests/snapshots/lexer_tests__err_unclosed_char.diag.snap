---
source: tests/lexer_tests.rs
expression: error_msg
---
[1;31m●[0m [1;31mE1000[0m: [1mInvalid token[0m [90m[Lexer][0m
[90m────────────────────────────────────────────────────────────[0m
  [1;34m→[0m [36m<input>[0m:[1;33m2[0m:[1;33m9[0m

  [90m   2[0m [90m│[0m
  [90m    [0m [90m│[0m let x = 'A
  [90m    [0m [90m│[0m         [1;31m^^^[0m
  [90m    [0m [90m│[0m

  [1;34mⓘ[0m Encountered an invalid token: ''A
'

  [33m💡[0m [97mCheck for unexpected characters or typos in your code[0m

  [34m📚[0m Learn more: [36mhttps://liva-lang.org/docs/errors/lexer#e1000[0m
[90m────────────────────────────────────────────────────────────[0m
